File: 0804_Unique_Morse_Code_Words/804.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int uniqueMorseRepresentations(vector<string>& words) {
        vector<string> table = {".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", 
                                "..", ".---", "-.-", ".-..", "--", "-.", "---", ".--.", 
                                "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", 
                                "-.--","--.."};
        set<string> diff;
        for (const string w: words) {
            string morse = "";
            for (const char ch: w) {
                morse += table[ch - 'a'];
            }
            diff.insert(morse);
        }

        // for (string str: diff)  
        //     cout << str << " ";
        return diff.size();
    }
};

int main() {
    vector<string> t1 = {"gin","zen","gig","msg"};
    Solution a;
    cout << a.uniqueMorseRepresentations(t1);
    return 0;
}

File: 0804_Unique_Morse_Code_Words/804.py
class Solution:
    def uniqueMorseRepresentations(self, words: list[str]) -> int:
        table = [".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", \
                 "..", ".---", "-.-", ".-..", "--", "-.", "---", ".--.", \
                 "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", \
                 "-.--","--.."]
        diff = set()
        for w in words:
            morse = ''
            for ch in w:
                morse += table[ord(ch) - ord('a')]
            diff.add(morse)
        return len(diff)
        
a = Solution()
print(a.uniqueMorseRepresentations(words = ["gin","zen","gig","msg"]))
print(a.uniqueMorseRepresentations(words = ["a"]))